### ISSUE TITLE
Per window etags call stack?

### ISSUE THREAD

**Issue by phye** [redguardtoo/emacs.d#746]
Hi,
I've been using emacs client to view source code of two different projects, say that one of them is in C++ and another in Go. 
As I've been tracking function definitions via `counsel-etags-find-tag-at-point`) in C++ project and then switched to another emacs client whose window contains golang code, hitting `C-t` accidentally pop the C++ function definition stack out and presents me with  C++ code in golang project. This is really not the desired behavior.

My question is, is it possible to have each window their own function definition/reference stack so that hitting `C-t` (=xref-pop-marker-stack=) won't affect behaviors in other windows?

**Comment by phye** [comment]
Not sure if you happen to check this, or if it's just too difficult to be implemented.
Also I would like to hear your comments on [projectile](https://github.com/bbatsov/projectile). I'm thinking of using this as the solution to the problem above but not sure if it's consistent with your config.

Thanks.

**Comment by redguardtoo** [comment]
I checked existing code. I already set `tags-file-name` as buffer local because when calling `visit-tags-table`, I set second parameter to `t`. Can you double check `tags-file-name` in two projects?

**Comment by phye** [comment]
Just tried in both projects using your config, found that `tags-file-name` in two projects are both nil. Is this expected?

**Comment by redguardtoo** [comment]
could be `nil`. but will be set by `vist-tags-table` after scanning code. let me have a look.

**Comment by redguardtoo** [comment]
check the new version `87f33b0 tags-file-name should bind to correct buffer (Chen Bin)` of `counsel-etags`

**Comment by phye** [comment]
Just merged your latest fix, still with no luck. 
`tags-file-name` is still nil.

**Comment by redguardtoo** [comment]
Yeah, I think I need a new plan. `tags-file-name` is not very helpful.
A global variable `counsel-etags-scan-history` is better. I need some time to figure out a better plan.

**Comment by phye** [comment]
Thanks for your time!

**Comment by redguardtoo** [comment]
check new version of `counsel-etags` (already included in this setup)

**Comment by phye** [comment]
Thanks for the effort!
But unfortunately the new solution seems not to work as well. 
Hitting `c-t` in c++ project will still pops golang call stack out.

**Comment by redguardtoo** [comment]
I fixed `C-]` and tags updates. You need press `C-t` at least once in new project.

`C-t` is command from `xref` (not written by me ). It reads from global variable `xref--mark-ring`. It's better not to respect Emacs default convention because other packages are using this global variable.

**Comment by karta0807913** [comment]
Hello, I wrote some code for this issue.

It works find, but I only test few case in my emacs.
```elisp
(setq xref-marker-ring-hash (make-hash-table
                             :data ("global" xref--marker-ring))) ;; save xref workspace

(evil-define-motion my-evil-goto-definition ()
  "Go to definition or first occurrence of symbol under point in current buffer."
  :jump t
  :type exclusive
  (let* ((string (evil-find-symbol t))
         (search (format "\\_<%s\\_>" (regexp-quote string)))
         (project-root (or (ffip-project-root) "global")) ;; find the project root in order to split the xref workspace
         ientry ipos)
    (if (not (gethash project-root xref-marker-ring-hash)) ;; init the value if key not exists
        (puthash project-root (make-ring xref-marker-ring-length) xref-marker-ring-hash))
    (make-local-variable 'xref--marker-ring) ;; make xref ring local
    (setq xref--marker-ring (gethash project-root xref-marker-ring-hash))

    ;; load imenu if available
    (my-ensure 'imenu)
    (evil-goto-definition) ;; goto the code

    (make-local-variable 'xref--marker-ring) ;; set target buffer xref to local
    (setq xref--marker-ring (gethash project-root xref-marker-ring-hash))
    ))
```

However, this code not supports the same upstream.
If you go to the same source code and backtrack(for example: main.c -> stdio.h and main.h -> stdio.h), it will brake the local variable xref--marker-ring in upstream (stdio.h).

![image](https://i.imgur.com/EA8kzy6.gif)
